metric/request.go

```diff
@@ -0,0 +1,79 @@
+package metric
+
+import (
+	"strconv"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/hermeznetwork/hermez-node/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	favicon = "/favicon.ico"
+)
+
+// Prometheus contains the metrics gathered by the instance and its path
+type Prometheus struct {
+	reqCnt *prometheus.CounterVec
+	reqDur *prometheus.HistogramVec
+}
+
+// NewPrometheus generates a new set of metrics with a certain subsystem name
+func NewPrometheus() (*Prometheus, error) {
+	reqCnt := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: namespaceAPI,
+			Name:      "requests_total",
+			Help:      "How many HTTP requests processed, partitioned by status code and HTTP method",
+		},
+		[]string{"code", "method", "path"},
+	)
+	if err := registerCollector(reqCnt); err != nil {
+		return nil, err
+	}
+	reqDur := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: namespaceAPI,
+			Name:      "request_duration_seconds",
+			Help:      "The HTTP request latencies in seconds",
+		},
+		[]string{"code", "method", "path"},
+	)
+	if err := registerCollector(reqDur); err != nil {
+		return nil, err
+	}
+	return &Prometheus{
+		reqCnt: reqCnt,
+		reqDur: reqDur,
+	}, nil
+}
+
+// PrometheusMiddleware creates the prometheus collector and
+// defines status handler function for the middleware
+func PrometheusMiddleware() gin.HandlerFunc {
+	p, err := NewPrometheus()
+	if err != nil {
```

Comment:
if `err != nil` then `p == nil`.  If `p == nil` the middleware will panic when dereferencing `p`, which is inconvenient, because it happens later on once an API request is done, it happens in a goroutine, gin-gonic may recover the panic, etc.  I suggest this function either returns error, or panics, instead of just logging the error.